src/Provider/IpFinder/Tests/IpFinderTests.php

```diff
@@ -0,0 +1,146 @@
+
+/*
+ * This file is part of the Geocoder package.
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @license    MIT License
+ */
+
+namespace Geocoder\Provider\IpFinder\Tests;
+
+use Geocoder\IntegrationTest\BaseTestCase;
+use Geocoder\Provider\IpFinder\IpFinder;
+use Geocoder\Query\GeocodeQuery;
+
+/**
+ * @author Jonas Gielen <gielenjonas@gmail.com>
+ */
+class IpFinderTests extends BaseTestCase
+{
+    protected function getCacheDir()
+    {
+        return __DIR__.'/.cached_responses';
+    }
+
+    public function testGetName()
+    {
+        $provider = new IpFinder($this->getMockedHttpClient());
+        $this->assertEquals('ipfinder', $provider->getName());
+    }
+
+    public function testGeocodeWithNoKey()
+    {
+        $provider = new IpFinder($this->getMockedHttpClient());
+        $this->assertEquals('free', $provider::DEFAULT_API_TOKEN);
+    }
+
+    public function testGeocodeKey()
+    {
+        $provider = new IpFinder($this->getMockedHttpClient(), 'TOKEN');
+        $this->assertEquals('TOKEN', $provider->apiKey);
+    }
+
+    /**
+     * @expectedException \Geocoder\Exception\UnsupportedOperation
+     * @expectedExceptionMessage The IpFinder provider support only IP addresses.
+     */
+    public function testGeocodeWithAddress()
+    {
+        $provider = new IpFinder($this->getMockedHttpClient());
+        $provider->geocodeQuery(GeocodeQuery::create('Egypt, France'));
+    }
+
+    /**
+     * @expectedException \Geocoder\Exception\InvalidArgument
+     * @expectedExceptionMessage No API Key was specified, invalid API Key.
+     */
+    public function testGeocodeWith401Code()
+    {
+        $json = <<<'JSON'
+{
+   "errors": [
+      {
+         "code": 401
+      }
+   ]
+}
+JSON;
+        $provider = new IpFinder($this->getMockedHttpClient($json), 'asdsad');
+        $result = $provider->geocodeQuery(GeocodeQuery::create('1.0.0.0'));
+    }
+
+    /**
+     * @expectedException \Geocoder\Exception\InvalidCredentials
+     * @expectedExceptionMessage The requested resource does not exist.
+     */
+    public function testGeocodeWith404Code()
+    {
+        $json = <<<'JSON'
+{
+   "errors": [
+      {
+         "code": 404
+      }
+   ]
+}
+JSON;
+        $provider = new IpFinder($this->getMockedHttpClient($json), 'asdsad');
+        $result = $provider->geocodeQuery(GeocodeQuery::create('1.0.0.0'));
+    }
+
+    /**
+     * @expectedException \Geocoder\Exception\InvalidCredentials
+     * @expectedExceptionMessage Method Not Allowed.
+     */
+    public function testGeocodeWith405Code()
+    {
+        $json = <<<'JSON'
+{
+   "errors": [
+      {
+         "code": 405
+      }
+   ]
+}
+JSON;
+        $provider = new IpFinder($this->getMockedHttpClient($json), 'asdsad');
+        $result = $provider->geocodeQuery(GeocodeQuery::create('1.0.0.0'));
+    }
+
+    public function testGeocodeWithIPv4()
```

Comment:
Can we have a few more example/tets?

Reply:
It seems the state is returned in the response but not set in this provider. Can you add that?